pkg/tsdb/azuremonitor/aztokenprovider/token_provider.go

```diff
@@ -180,6 +215,119 @@ func (c *clientSecretTokenRetriever) GetAccessToken(ctx context.Context, scopes
 	return &AccessToken{Token: accessToken.Token, ExpiresOn: accessToken.ExpiresOn}, nil
 }
 
+type userIdentityTokenRetriever struct {
+	tokenEndpoint string // token endpoint to retrieve the token from
+	authHeader    string // exact Authorization header to be used to talk to the token endpoint
+	client        *http.Client
+}
+
+func (c *userIdentityTokenRetriever) GetCacheKey() string {
+	return fmt.Sprintf("azure|useridtoken|%s", c.tokenEndpoint)
```

Comment:
Cache key should include user id (or any user discriminator), so different users wouldn't reuse the same token, and token endpoint isn't actually needed in the cache key because it's the Grafana instance configuration which is static.